nexus/db-queries/src/db/datastore/snapshot_replacement.rs

```diff
@@ -0,0 +1,1428 @@
+
+                async move {
+                    use db::schema::snapshot_replacement_step::dsl;
+                    use db::schema::volume_repair::dsl as volume_repair_dsl;
+
+                    // Skip inserting this new record if we found another
+                    // snapshot replacement step with this volume in the step's
+                    // `old_snapshot_volume_id`, as that means we're duplicating
+                    // the replacement work: that volume will be garbage
+                    // collected later. There's a unique index that will prevent
+                    // the same step being inserted with the same volume id.
+
+                    let maybe_record = dsl::snapshot_replacement_step
+                        .filter(dsl::old_snapshot_volume_id.eq(request.volume_id))
+                        .get_result_async::<SnapshotReplacementStep>(
+                            &conn,
+                        )
+                        .await
+                        .optional()?;
+
+                    if let Some(found_record) = maybe_record {
+                        return Err(err.bail(Error::conflict(format!(
+                            "{:?} already referenced in old snapshot volume for \
+                            request {:?}",
+                            request.volume_id,
+                            found_record.id,
+                        ))));
+                    }
+
+                    // The snapshot replacement step saga could invoke a volume
+                    // replacement: create an associated volume repair record.
+
+                    diesel::insert_into(volume_repair_dsl::volume_repair)
+                        .values(VolumeRepair { volume_id: request.volume_id, repair_id: request.id })
+                        .execute_async(&conn)
+                        .await?;
+
+                    diesel::insert_into(dsl::snapshot_replacement_step)
+                        .values(request)
+                        .execute_async(&conn)
+                        .await?;
+
+                    Ok(())
+                }
+            })
+            .await
+            .map_err(|e| {
+                if let Some(err) = err.take() {
+                    return err;
+                }
+
+                public_error_from_diesel(e, ErrorHandler::Server)
+            })
+    }
+
+    pub async fn get_snapshot_replacement_step_by_id(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+    ) -> Result<SnapshotReplacementStep, Error> {
+        use db::schema::snapshot_replacement_step::dsl;
+
+        dsl::snapshot_replacement_step
+            .filter(dsl::id.eq(id))
+            .get_result_async::<SnapshotReplacementStep>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    pub async fn get_requested_snapshot_replacement_steps(
```

Comment:
The naming here has me a little confused.

We are getting all the `snapshot_replacement_step`s that are in state `Requested`?

Reply:
Correct yeah